### ISSUE TITLE
Idea: have a _view_count for all widgets

### ISSUE THREAD

**Issue by maartenbreddels** [jupyter-widgets/ipywidgets#1225]
For performance reasons I need to know the number of views. I know @SylvainCorlay doesn't like exposing this. But let me give some use case(s).

Say I have at Output widget that will catch the output of matplotlib, that is the result of a really computational expensive result (yeah, vaex). Some external event, say a slider, or anything else can trigger a redraw of this plot, which might take several seconds. 
Now I reexecute the cell, so that the previous Output widget (or actually its view) is gone, although I still have a reference to it somewhere. Now the some event will trigger a redraw of both Output widgets, so it will take ~5 second. 1000 cell executions down the road, I have 500 output widgets, that will all take update a matplotlib figure, that is not visible, not good.

If however, we know a widget has 0 views, we know we don't have to do the lengthy calculation, and possible, when an old widget is 'printed' again, its view count jumps from 0 to 1 and we know we need to update it.

TLDR: Would be really useful to know if a widget is visible (_view_count property) so we can skip any unnecessary update.

**Comment by jasongrout** [comment]
I think it makes sense to have some way for a model to know if it can optimize based on the views it is feeding.

**Comment by maartenbreddels** [comment]
Ok, I might do a PR on this then.

**Comment by jasongrout** [comment]
Just had another thought - Suppose widget A is listening to changes on widget B, and widget B is doing some sort of optimization like this - not updating and sending changes to the frontend. Now widget A will also go stale, unless it has some way of telling B that the updates are needed. In other words, B can't assume that its views are the only things listening to the model changes.

Granted in the vast majority of times, this won't happen, and the view count will be sufficient. However, do you have an answer for the times when one object is listening to changes in the other object?

**Comment by maartenbreddels** [comment]
I'd say we should not cover those use cases (yet), to keep it simple. For instance, if I plan to use _view_count for this strategy in vaex, I should document that I do it that way, and possibly have an option to avoid this behaviour. 

Just to be clear, the 'going stale' is a implementation detail by 3rd parties, not something ipywidgets would do nor recommend.

**Comment by jasongrout** [comment]
More generally, it would be great to have a more reactive system, where the end consumers (here notably the views, but could be also other things listening to change requests) request changes and the updates are made lazily.

For example, another clear optimization is a view that realizes it's not in the current viewport, and so can be updated lazily. If we instead had a system for views requesting updates, it could solve both optimizations in one stroke.

Perhaps something that is closer to the semantic meaning we want here is that a particular model has an 'active' attribute, and the model on the javascript side sets this to true when it knows the model should be updating and pushing data.

**Comment by jasongrout** [comment]
Hmmm...thinking of the information flow, this is also now the only attribute in the core widget model that is completely controlled by the frontend. Which means that if a frontend disconnects for some reason, the view count may not be updated and may be false.

**Comment by maartenbreddels** [comment]
about the view_count not being updates, yes, I'd say it needs to be set to zero on a disconnect, do you know how we could handle that? 
For the lazy thing, I agree, that would be ideal, maybe we should do it in small steps, for instance, adding a property that says if a view is on screen would be the next logical step

**Comment by jasongrout** [comment]
Fixed in #1232.
